### ISSUE TITLE
MCTS alternatives

### ISSUE THREAD

**Issue by roy7** [gcp/leela-zero#860]
Since a lot of people are working on tuning FPU at the moment and some people are exploring tweaks to the search algorithm I wanted to share a few areas of research I was looking over this evening, in case the more mathematical programmers among us were interested in trying these with LZ.

After reviewing [Multi Armed Bandits and Exploration Strategies](https://sudeepraja.github.io/Bandits/) I was interested in the mention of Thompson Sampling. It's an alternative to UCB that also has a logarithmic bound on regret and has been known since the 1930s, but the theoretical groundwork on it has only recently been finished in 2012 in the paper [Analysis of Thompson Sampling for the Multi-armed Bandit Problem](http://proceedings.mlr.press/v23/agrawal12/agrawal12.pdf). That paper also covers the stochastic case, which makes it more applicable for us by people like us.

So I searched around some more if anyone has used this with MCTS yet, and it turns out this is a pretty hot current research area. There's a 2014 paper [Thompson Sampling Based Monte-Carlo Planning in POMDPs](https://aijunbai.github.io/publications/ICAPS14-Bai.pdf), and two 2017 master's thesis I found, [Thompson Sampling for Monte Carlo Tree Search and Maxi-min Action Identification](https://www.universiteitleiden.nl/binaries/content/assets/science/mi/scripties/master/2017-2018/master-thesis_mingxi-li_12okt2017.pdf) and [Monte Carlo Tree Search with Thompson Sampling in the Settlers of Catan](http://eprints.fri.uni-lj.si/3504/1/63100285-Katja_Tuma-Drevesno_preiskovanje_Monte_Carlo_s_Thompsonovim_vzor%C4%8Denjem_pri_igri_Prebivalci_otoka_Catan_(3).pdf) (two languages). 

A recent paper following up on these things is [Enhancements in Monte Carlo Tree Search
Algorithms for Biased Game Trees](https://pdfs.semanticscholar.org/4526/21321d943c924553fc2ad2e8e12b2225d7ea.pdf). This paper discusses a way to deal with biases in game trees that might be useful for us, applying to 3 types of MCTS (UCT, TS, and KL-UCB). I hadn't heard of KL-UCB before, this paper mentioned that it performs better and has better theoretical bounds on regret than UCT does. The source for KL-UCB is the paper [KULLBACK–LEIBLER UPPER CONFIDENCE BOUNDS FOR OPTIMAL SEQUENTIAL ALLOCATION](https://pdfs.semanticscholar.org/4526/21321d943c924553fc2ad2e8e12b2225d7ea.pdf).

So for those interested in recent research, you might find some of these interesting. I don't know of any Go projects that use anything other than UCT for MCTS, I imagine most research effort goes into the neural network side of things and not improving the tree search side of things. Perhaps this is an area someone would like to do some original research into.

**Comment by Eddh** [comment]
Seems interesting. It'll take time to read and understand all that though. I'm not that good at math.

**Comment by wpstmxhs** [comment]
We may achieve what AlphaGo team couldn't do. Maybe they didn't think about changing tree search algorithm, due to lack of useful alternatives on that time.

**Comment by wpstmxhs** [comment]
For your reference, there is an interesting experiment from japanese computer go developer board, done by HiraBot developer. he used negamax algorithm instead of MCTS, and saw it seems working not bad with even small thinking time.

http://524.teacup.com/yss/bbs/?page=13&

**Comment by roy7** [comment]
Here's a detailed overview that covers a wide variety of things I'm unfamiliar with.

[A Survey of Online Experiment Design with the Stochastic Multi-Armed Bandit](https://arxiv.org/pdf/1510.00757.pdf).

**Comment by barrtgt** [comment]
This paper is also interesting: http://ieeexplore.ieee.org/document/7352306/
It was cited from the zero paper. They optimized the search parameters through a Gaussian process using data from a preliminary run.

**Comment by gcp** [comment]
Optimizing the search parameters is not a problem. CLOP is very good at that. Any paper that doesn't compare to it can safely be ignored.

**Comment by odeint** [comment]
The CLOP book chapter is here by the way:
https://link.springer.com/chapter/10.1007/978-3-642-31866-5_13
Free version:
https://www.remi-coulom.fr/CLOP/CLOP.pdf

This was before neural networks, but is a general tuning algorithm.
It doesn't have that many citations though:
https://scholar.google.ch/scholar?cites=6152564073462446443&as_sdt=2005&sciodt=0,5&hl=en

There is this section in the AlphaZero paper ( https://arxiv.org/abs/1712.01815 ):
> AlphaGo Zero tuned the hyper-parameter of its search by Bayesian optimisation. In AlphaZero
we reuse the same hyper-parameters for all games without game-specific tuning. The
sole exception is the noise that is added to the prior policy to ensure exploration (29); this is
scaled in proportion to the typical number of legal moves for that game type.

I'm not sure what exactly that Bayesian optimisation refers to. In the actual AlphaGo Zero paper ( https://www.nature.com/articles/nature24270 / free https://deepmind.com/documents/119/agz_unformatted_nature.pdf ), the only tuneable parameter mentioned is c_puct. Tuning a single parameter doesn't seem particularly involved or require special techniques:

> [...] where c_puct is a constant determining the level of exploration; this search control strategy initially
prefers actions with high prior probability and low visit count, but asympotically prefers actions
with high action-value

What would one tune in LeelaZero using CLOP? Puct for sure, but what else? Some of the newly introduced variables like regret, risk, opportunity, FPU reduction? That would still be very ad-hoc even if the tuning is rigorous, since the variables themselves were just 'invented' by fiat.

**Comment by gcp** [comment]
> What would one tune in LeelaZero using CLOP? Puct for sure, but what else?

Right now only CLOP and (if appropriate) FPU reduction.

>That would still be very ad-hoc even if the tuning is rigorous, since the variables themselves were just 'invented' by fiat.

I have no idea what that is supposed to mean.

>Tuning a single parameter doesn't seem particularly involved or require special techniques:

I wonder if there is no confusion there between the hyperparameters of the training run and those of the search.

**Comment by kiudee** [comment]
@roy7 I did a lot of experiments with Thompson sampling and other bandit algorithms for my Bachelor’s and Master’s thesis.
In a nutshell, Thompson sampling outperforms other stochastic bandit algorithms (KL-UCB, BayesUCB, UCB, ε-greedy etc) on most bandit problems.

Implementing it for classical MCTS is as simple as initializing a [Beta distribution](https://en.wikipedia.org/wiki/Beta_distribution) during node creation and updating it by using +wins and -losses. A decision on which action to take can be obtained by randomly sampling from each distribution and taking the move with the highest value.

It is more difficult if you do not evaluate rollouts until the end of the game, because we get the evaluations from a neural network. Thus you lose the [conjugacy](https://en.wikipedia.org/wiki/Conjugate_prior) of the Beta distribution and have to fit another distribution to the values returned during tree search. One could for example think of a normal distribution with certain priors, but one always has to consider the performance of the whole fitting process (remember for Thompson sampling the most important operation is _sampling_ from the distribution).

**Comment by odeint** [comment]
> I wonder if there is no confusion there between the hyperparameters of the training run and those of the search.

Aren't the two entangled, because the networks are trained to predict the outcome of the search? Is it possible that the search algorithm which delivers the highest playing strength for a given network is actually not the best tutor for the next generation of networks?

**Comment by kiudee** [comment]
@barrtgt @odeint Bayesian optimization refers to fitting a stochastic response surface model to the outcome of an experiment (usually the performance metric of a machine learning model). [scikit-optimize](https://scikit-optimize.github.io/) for example is a Python package which implements that.
This usually works best for real-valued outcomes, since Gaussian processes fit a smooth function.

It is possible to fit it on binary outcomes (e.g. Go matches), but for that you need to fit a latent variable model where you have to integrate over the binary squashing function. For that you usually need MCMC machinery. For binary outcomes one therefore usually applies CLOP or SPSA.

**Comment by killerducky** [comment]
@kiudee 
> It is more difficult if you do not evaluate rollouts until the end of the game, because we get the evaluations from a neural network. Thus you lose the conjugacy of the Beta distribution and have to fit another distribution to the values returned during tree search.

Why does it matter we use the NN instead of rollouts? I thought a Beta distribution can model the NN value outputs, which are real values in the range [0,1].

**Comment by kiudee** [comment]
@killerducky When using binary outcomes the parameters α and β can be updated directly. For example: If a subtree won 10 times and lost 3 times you can directly sample from the posterior distribution `Beta(1+10, 1+3)` (the initial ones constitute a uniform prior distribution over the winrate).
For real-valued outputs in [0,1] you would have to fit the parameters using maximum likelihood or similar inference methods.

**Comment by roy7** [comment]
The interesting thing the authors of the Thompson sampling regret bounds paper do for the stochastic case is randomly choose a win or a loss by the [0,1] result observed and store that as a win or a loss instead of using the value directly. It feels like that's throwing away info, yet it works in practice apparently.

The problem with tree search is as new moves are explored and we realize the real value of the board is much different than the earlier NN eval thought it was, the win rates of the node can change a lot and basically be a different distribution than we were sampling from before.

**Comment by wctgit** [comment]
@roy7 In the MCMC world, that's referred to as the 'burn in' period. Basically, they spend lots and lots of cycles on finding the approximation to the posterior probability distribution, whereas we are generally trying to build a system that has massive constraints on the number of cycles we can 'burn' to determine the posterior, i.e. the best move. We are trying to find a 'pretty good' approximation ASAP, so we are inherently going to be running into this problem with not-enough-burn-in all the time, IMHO.

**Comment by kiudee** [comment]
@roy7 Indeed, that is because we assume a fixed distribution over time.
One could circumvent this problem by modeling each result to be generated by an adversary and use adversarial bandit algorithms (e.g. Exp3) to solve it. But these algorithms have very slow convergence due to the need to be robust against the adversary. There was [work done by Sébastien Bubeck and Aleksandrs Slivkins (2012)](http://proceedings.mlr.press/v23/bubeck12b/bubeck12b.pdf) on finding good hybrid algorithms. [Seldin and Slivkins (2014)](http://proceedings.mlr.press/v32/seldinb14.html) improve on that by proposing the algorithm Exp3++.
Here is the Figure from the paper and you can see that there is still quite a big gap between Exp3++ and Thompson Sampling when applied on _stochastic_ data:
![screenshot-2018-2-14 one practical algorithm for both stochastic and adversarial bandits - seldinb14 pdf](https://user-images.githubusercontent.com/564473/36211289-1f0d4f38-11a1-11e8-8860-8205acf172f8.png)
How it works out in MCTS needs to be determined though.
[Seldin and Lugosi (2017)](https://arxiv.org/pdf/1702.06103.pdf) further optimized the parameters of Exp3++, but did not perform empirical comparisons.

[Auer and Chiang (2016)](http://www.jmlr.org/proceedings/papers/v49/auer16.pdf) provide a recent improvement on the pseudo regret, but without empirical comparison.

TL;DR: We should do experiments on which bandit algorithm improves the _match results_ the most. This already incorporates the disadvantage caused by a slower bandit algorithm.

**Comment by roy7** [comment]
@kiudee An idea I want to toy with once I figure out how in the code is to adjust the values of nodes to not be the average of all of the values under the node, but only the average of values under the node where the child node averages are within an overlapping confidence interval of the best moves. This is motivated by the other discussion https://github.com/gcp/leela-zero/issues/855 with the MV% stat showing an obvious warning about a blunder that was made, and the traps not discovered until deep in the search only slowly lower the win rate and the blunder move can't be avoided until a different move has more total visits.

In a nutshell, consider where a blunder is about to be made and opponent can do a winning move next turn but I don't see it at first. So my move is AA and the opponent has two choices, LL for opponent to lose, and WW to win. LL is much more likely to be searched from my heatmap so at first I'm really confident I'm winning. Eventually I'm at:

1000 visits AA LL (90% win rate for me)
50 visits AA WW (15% win rate for me)

The bounds on that 50 might be pretty wide since it's only 50 visits, but for now I see this move as about a 86% win rate move. Since this is the opponent's move they want to minimize my win rate, so future search goes mostly into AA WW. But once we're to this point:

1000 visits AA LL (90%)
250 visits AA WW (15%)

I probably have enough visits to realize the confidence interval around the 90% and the 15% are not touching any longer. In that case, instead of calling this node's value 75% I should call it 15%. I have enough certainty that the inferior move for the opponent is not the same distribution as the superior move, so the inferior move will not be taken and so it should just be ignored.

Sort of like mixmax, but not strictly using the value of the best/worst node. Using normal MCTS methodology... except nodes we know with some confidence aren't similar to the best choice for that player at that time won't skew the score up/down.

**Comment by wctgit** [comment]
Looks like a reasonable idea. A part of me wants to play devil's advocate, though, and say, "This would just be an example where the net needs more training, so that it sees AA WW as higher probability. Then it the MCTS would automatically already converge to the right answer more quickly, guided by the smarter net. Should we really be messing around so much with the way the MCTS performs searches, when really we just need to train the net more?" But then again, I can see an argument that supports your proposal, which is that this particular assumption of having an opponent who wants to minimize your score is not specific to Go, and so the minimax utility of such a tweak seems justified in the general case. (Just some random thoughts.)

**Comment by kiudee** [comment]
@roy7 Yes, this problem can be circumvented by selecting the move with the highest  lower confidence bound (or quantile if you are in the Bayesian setting).

Note, how this is exactly the reverse methodology to what we are doing during search. There we want to be _optimistic_ in order to balance exploration/exploitation. When selecting the moves we want to be _pessimistic_ since we still have uncertainty in each estimate which we want to account for. Selecting based on visits only makes sense if the bandit algorithm has converged.

**Comment by jkiliani** [comment]
@kiudee Sounds very logical, but how would you change training data if move selection was based on highest lower confidence bound, instead of visit count? At the moment, using visit counts for training is very logical, but with this move selection method there should also be a better way...

**Comment by kiudee** [comment]
Good question, for the normal search one really should use optimistic methods (UCB, Thompson Sampling etc), because you otherwise miss even more moves. The pessimistic aspect is only useful when you have to make a move in the end.

For final move selection one could think about running a shallow "minimax" like search of applying the LCB until depth _k_. To avoid situations @roy7 talked about (where actually have the correct knowledge already).

**Comment by jkiliani** [comment]
In general, would it likely benefit the search if policy priors were a bit broader (i.e. less lopsided toward what the search determined was the best move)? Would something like exp(-c*LCB) with a slope constant still to be determined work well as a policy training target? Or even much simpler, would it benefit the search if the policy targets were simply flattened by applying something like the square root to them?

**Comment by kiudee** [comment]
@jkiliani That is actually what the adversarial bandit algorithms are doing. The rewards are exponentially weighted and uniform noise is mixed in to be robust. [Here](https://jeremykun.com/2013/11/08/adversarial-bandits-and-the-exp3-algorithm/) is a good overview.

**Comment by gcp** [comment]
>In general, would it likely benefit the search if policy priors were a bit broader (i.e. less lopsided toward what the search determined was the best move)?

From experience, no, it's the other way around. You can also see this in the original AGZ paper, where the optimal softmax_temp = 0.67, i.e. making the probabilities more extreme.

**Comment by jkiliani** [comment]
> From experience, no, it's the other way around. You can also see this in the original AGZ paper, where the optimal softmax_temp = 0.67, i.e. making the probabilities more extreme.

OK, so flattening doesn't work. What about using LCB for move selection and training targets, as the link posted by @kiudee describes? One thing that always bugged me about using simple visit count for move selection and training is the situation when the search has already determined that a more recently explored move is actually better, but there are not enough playouts left to overtake the PV.

**Comment by gcp** [comment]
Theorizing about whether it would work is pretty pointless IMHO. Just implement and test it.

**Comment by gcp** [comment]
I mean, implement LCB for move selection and see what it does for the playing strength. That's the first step. What to do with the training targets is trickier.

**Comment by WhiteHalmos** [comment]
I think it's definitely worth testing just improving the move selection. My resources are tied up in FPU testing, but for those who want to make this change and test it out, please test at least 100 games. Thank you!

**Comment by roy7** [comment]
> I mean, implement LCB for move selection and see what it does for the playing strength. That's the first step. What to do with the training targets is trickier (not sure they need changing? as the tree search still uses UCB).

I do hope to implement this myself. I just wanted @kiudee's feedback though since he's an expert in this stuff and perhaps could give me some guidance before I started. :) It sounds like instead of my initial idea of only averaging the nodes with overlapping confidence intervals (in other words, subtract out ones we know with statistical significance are worse moves we will not choose) what I should explore is taking the final resulting search tree and score it out based on highest LCB in terms of which move to actually take. That should at least avoid immediate blunders we discovered late in the search.

This was kind of like the MV% display, but I believe that just used a pure minmax idea of "worst winrate of all opponent's choices" without calculating the confidence intervals.

Later work could include my initial idea during the search itself (so a dead end node doesn't need the win % to slowly drop over time, we can drop it earlier if we know the prior good results are no longer going to happen if the good results are on a node whose confidence bounds have diverged from the best move).

**Comment by WhiteHalmos** [comment]
+1, this is an area ripe for optimizing, since the tree search was part of the AGZ paper but the move selection was just a simple max on top of the tree search.

**Comment by jkiliani** [comment]
Should the confidence intervals at root be calculated with the same cfg_puct as in the node expansion? Are there any reasons why this might not work well?

**Comment by kiudee** [comment]
@roy7 Sounds good. The simplest thing to try is selecting by highest LCB in the root node and compare it to the current move selection. If that works a depth 1 minimax LCB could be tried.

For the bandit algorithms if you want to try Thompson Sampling we would need an implementation for the Beta distribution in C++ which supports fast fitting and sampling. Alternative would be to use a normal distribution with conjugate prior distribution (Normal-InverseGamma-Model) which could be faster to update (and the Beta distribution approaches the Normal distribution in the limit).

**Comment by roy7** [comment]
Ok my first baby steps in this direction are in. I don't know why I thought gomill/ringmaster did SPRT tests and would stop early. It'll go all the way to the 400 I specified right? I used that instead of validation this time, so I'm unsure when to stop, hah.

Manually checking it seems I'm close a pass but not quite yet.

This is for the simplest use of LCB in move selection. No tree search, and no change to the search algorithm (unless sorting nodes changes those things). I changed sort_children to sort by lower confidence bounds as the first step, falling back to visits/etc in case of ties.

Options were "--timemanage off -r 1 --playouts 1600 -w w  --noponder --gtp" using network 8f00.

```
next v max_lcb (163/400 games)
board size: 19   komi: 7.5
          wins              black         white       avg cpu
next        67 41.10%       35 40.23%     32 42.11%    182.30
max_lcb     96 58.90%       44 57.89%     52 59.77%    191.11
                            79 48.47%     84 51.53%

```

**Comment by alreadydone** [comment]
A new paper from Alberta; Martin Mueller in the authors. Specifically mentions Go.
http://webdocs.cs.ualberta.ca/~mmueller/ps/2018/Chenjun-Xiao-M-MCTS-aaai18-final.pdf

**Comment by Eddh** [comment]
@roy7 nice. Now the question that will be harder to answer is, will it be good in training ? The policy will have to learn something that is only partially related to the move that is played in the end. I guess this is bad. But maybe it is ofset by the better quality games ?

**Comment by roy7** [comment]
@Eddh Well with this first step no change to training is needed at all. The existing training system can keep doing optimistic search and target the network to learn those priors. Although making stronger move selection does mean the games might go in directions they wouldn't have otherwise. But the training targets don't need to change?

Even if I come up with a way to change how nodes are valued to alter the search algorithm, ultimately training wouldn't need to change for that either. New values would shape search differently, but those new search priors would get learned as normal.

Where training would need to change is if we did something drastic (which we won't as part of the current LZ project) to, say, learn node lower confidence bounds instead of learning the % of search that will get allocated to that move. Although this might be harder to learn or unstable. Right now learning a move in a given board state is the best move and should get 95% of the search, it doesn't matter what the value network says as long as this move is still the relative best move. But if we're learning something like the LCB for every node instead, that'll change every time our value network changes. Might be harder to learn, I don't know.

**Comment by roy7** [comment]
Running my lcb at root only test on an old weak network now (1ac2). This will take a while, it's 23-19 so won't be an early SPRT pass and I suspect these games run longer since the weak network barely knows how to win.

**Comment by WhiteHalmos** [comment]
@killerducky linked to this on reddit [Learning to Search with MCTSnets](https://arxiv.org/pdf/1802.04697.pdf). Perhaps our days of twiddling fpu has finally come to an end. 😉

**Comment by Eddh** [comment]
@WhiteHalmos we would need to start a go ai from scratch though.

**Comment by WhiteHalmos** [comment]
Maybe testing it out with 7x7 or 9x9 go would be feasible by a single person with a decent machine and lots of time. 😄

**Comment by jkiliani** [comment]
I'm a bit unclear about the implications here: Could the nets they describe in this paper be trained just once, with a single reasonably strong LZ-Network, similar to an MCTS parameter tuning? Or would this require training the actual Go-playing net from scratch as well?

**Comment by roy7** [comment]
Wanted to store this someplace useful.

[Confidence Bound Algorithms in Game Trees](https://www.universiteitleiden.nl/binaries/content/assets/science/mi/scripties/statscience/mvolkert_thesis_19okt2017.pdf)

**Comment by roy7** [comment]
[Bayesian Inference in Monte-Carlo Tree Search](https://pdfs.semanticscholar.org/a37f/480cf340b6cd2c2fda4db31e7cc89bcddc8d.pdf)

**Comment by roy7** [comment]
Reading Thompson Sampling Based Monte-Carlo Planning in POMDPs more this afternoon, I think that might be usable for LZ. Whether or not I can implement it (or will even try) is the question. 

What was throwing me off on some Thompson Sampling approaches is when they would model the probabilities as Beta distributions with something like B(1+wins, 1+losses). Of course that only works for success/fail results, not stochastic results.

So then I saw a solution someone proposed to take a stochastic result and just use a 0 or a 1 with those odds each time. But that felt like throwing away a lot of information, and we aren't pulling a bandit arm over and over. And we only evaluate each node once.

The above mentioned paper seems to work better for what we'd do, and although they talk about generating a value of a leaf using rollouts, is isn't fundamental to what is happening with the algorithm. The rollout just generates a reward that gets passed back up the tree, like we do already (using the nn_eval as the reward).

**Comment by tterava** [comment]
I've been testing out your pull request #883 and it seems to make a huge difference at low playouts. LZ is a lot faster to switch to a more optimal move if it happens to find one late in the search. With slower time control it doesn't seem to make that much difference, but another approach might yield better results. It seems there's a lot of gains to be had with move selection. Selecting based on visits is too simplistic and definitely isn't the best approach overall.

**Comment by Pytq** [comment]
@roy7 if you get value v_1, ..., v_n couldn't you use B(a+b*sum(v_i), a+b*sum(1-v_i)) for some a and b?
This way you still have your distribution concentrating around the emperical mean of v_1, ..., v_n as you get more values and you use all information available.

**Comment by tterava** [comment]
@roy7 

> The above mentioned paper seems to work better for what we'd do, and although they talk about generating a value of a leaf using rollouts, is isn't fundamental to what is happening with the algorithm.

Which of the papers are you referring to? I'm interested in trying to understand the different approaches a bit better and I might even have time to try and implement some of them.

**Comment by roy7** [comment]
@tterava Check the link earlier for "Thompson Sampling Based Monte-Carlo Planning in POMDPs". That's the one I was referring to. :)

**Comment by gcp** [comment]
>What was throwing me off on some Thompson Sampling approaches is when they would model the probabilities as Beta distributions with something like B(1+wins, 1+losses). Of course that only works for success/fail results, not stochastic results.

Why wouldn't it work with stochastic results? beta is a continuous distribution, so sampling from B(1+0.5, 1+0.5) just works.

Thompson sampling has some interesting properties over UCB1, i.e. it is fully random, which means it parallelizes better, and it deals with FPU issues in a much better way because you can set sensible priors for everything.

But there are also some very thorny issues. Like how do you incorporate the policy prior into it (which is required for Zero to work)? There's no computationally cheap way that I can think of to mix the beta distributions with a policy prior.

**Comment by kiudee** [comment]
If you want to preserve the conjugacy property to easily update the
distribution in each node, one option is the normal inverse-gamma
distribution. It's support is not restricted to (0,1), but that is not a
problem, because we only care about the relative order.
A prior can be centered on the policy.

For Thompson sampling it's important, that you find a fast implementation
for the sampling.

Gian-Carlo Pascutto <notifications@github.com> schrieb am Do., 19. Apr.
2018, 20:49:

**Comment by roy7** [comment]
I guess what threw me off was one of the papers I read mentioned handling stochastic results by returning a 1 or a 0 with the stocastic result as the random chance of which to return. I don't recall which paper it was now though, or reasoning to do that vs just call it a .8 win and .2 loss.

**Comment by gcp** [comment]
>There is a way to still use the beta distribution, by throwing a biased coin to make the outcome binary. Of course that throws away a lot of information.

This is what jroy is saying, but it's obviously suboptimal, and as far as I understand, not actually needed? (See my post)

**Comment by gcp** [comment]
>If you want to preserve the conjugacy property to easily update the distribution in each node, one option is the normal inverse-gamma distribution. It's support is not restricted to (0,1), but that is not a problem, because we only care about the relative order. A prior can be centered on the policy.

Can you explain a bit more how you believe this should work? inverse-gamma means we model the outcomes as a normal distribution instead of a bernouilly one, right? That may be more appropriate to neural net evaluations. But I think in that case we have to start tracking variance?

**Comment by kiudee** [comment]
In Thompson sampling we model the posterior distribution of the _mean_ reward.

Clarification
-------------
The beta distribution is the conjugate prior distribution for a Bernoulli/Binomial distribution. The signal/data is binary (win/loss), but the parameter μ being estimated is the mean of the data distribution (the biasedness of the coin).
Conjugacy here means that the posterior, after updating the prior with new data, is also a beta distribution.

Now, if our rewards are no longer binary, we can’t use the beta distribution anymore, since it’s parameters are wins/losses.
Our data is now the real valued evaluation in [0, 1] we observe from a playout.
The parameter μ we want to estimate is also a real value in [0, 1].

Considerations
--------------

We want to avoid fitting a distribution with some iterative maximum likelihood or Markov chain Monte Carlo method, because it is simply too much computation in each node. The conjugacy of the beta distribution was very nice. Thus it makes sense to look for a distribution of rewards we can use, that has that property.

In the [Normal-gamma distribution](https://en.wikipedia.org/wiki/Normal-gamma_distribution#Posterior_distribution_of_the_parameters) (or reparameterized Normal inverse-gamma) the assumption is as that the rewards (evaluations from playouts) are normally distributed.
Since we do not know either the mean μ or the standard deviation σ, we will assume a prior distribution for both. The Normal-gamma distribution is the prior distribution which achieves conjugacy, and it can be updated as follows:
If our prior is:
![screenshot-2018-4-20 normal-gamma distribution - wikipedia 1](https://user-images.githubusercontent.com/564473/39037514-0450b6d6-4481-11e8-8cab-50c508f6c4b9.png)
then the posterior distribution becomes (after _n_ observations):
![screenshot-2018-4-20 normal-gamma distribution - wikipedia](https://user-images.githubusercontent.com/564473/39037182-02010b02-4480-11e8-8782-f6acb817a6ea.png)
Where μ_0 denotes the prior mean of the distribution and can be initialized to the policy prior. Similar to the beta distribution, where we add "pseudo-wins/losses", λ_0 does the same here for the Normal-gamma distribution.
The prior parameters of the gamma distribution α_0, β_0 can be estimated once and then stay fixed. Maybe we could even tune them - I am not sure there.

**Comment by gcp** [comment]
I know all what you say but it doesn't address the critical part:

>Where μ_0 denotes the prior mean of the distribution and can be initialized to the policy prior.

What? I don't see how this is possible. Again, policy prior are a probability distribution over the moves, not over the outcomes.

**Comment by kiudee** [comment]
Sorry, my mistake. I meant the current eval we have for the parent. The exact value is not important, since we quickly override it with observations.

**Comment by gcp** [comment]
Right but then you still have no way to incorporate the policy prior, which is essential for the learning cycle to work. I also doubt your algorithm can be competitive if that knowledge isn't accounted for.

**Comment by gcp** [comment]
>we can’t use the beta distribution anymore, since it’s parameters are wins/losses.

That's not really true. Sure, you can interpret them as such, but:
https://stats.stackexchange.com/questions/105370/properly-interpret-the-alpha-beta-parameters-in-the-beta-distribution

**Comment by gcp** [comment]
>Regarding fpu

FPU isn't a problem in Thompson sampling. It's an issue specific to the UCB1 formulation which requires you to specify a scalar winrate for all nodes, and whatever value you pick is bad in some way. But in Thompson sampling you only need to specify a prior distribution.

**Comment by kiudee** [comment]
What about parent_eval * policy_prior as initialization. That distributes the parent eval to each node. It is not the true value, but without evaluating the value network on each node, we can’t obtain that.
The only point is anyway to focus on promising nodes first.

**Comment by Eddh** [comment]
> If you look at the modified UCB1 algorithm that was used in AZ, it scales the entire exploration term by the policy prior. So it's not just about early on in the search. It's a difference from the literature, which does use the "prior" in a manner that fades it out with a lot of playouts.

That's not completely true. They do say explicitely in the paper that is meant to focus on high policy first and converge on high winrate after. It's true that their formula is a far more long term prior than others that exist in the literature but it does fade over time.

**Comment by kiudee** [comment]
> That's not really true. Sure, you can interpret them as such, but:
stats.stackexchange.com/questions/105370/properly-interpret-the-alpha-beta-parameters-in-the-beta-distribution

True, you can [reparametrize the Beta distribution](https://en.wikipedia.org/wiki/Beta_distribution#Two_parameters_2) to using mean and sample size, which might be more useful in this case.

**Comment by roy7** [comment]
The Beta in some of these papers just seems so clean to use and understand for someone with limited math expertise like me. (I did minor in math during my B.S. but that 25 years ago.)

The [paper I was reading earlier](https://aijunbai.github.io/publications/ICAPS14-Bai.pdf) takes a further step in complexity:

> We perform statistical inference on the posterior distribution in Bayesian settings by choosing the conjugate prior in the form of a combination of two Dirichlet distributions and a NormalGamma distribution.

First step though is just trying out a basic UCT1 replacement (KL-UCB, Thompson Sampling, whatever) and getting that working in an expected way and seeing if we get at least similar results to the UCT1 search (if not better). I'm glad people who understand this better than me are taking an interest. I do believe stronger search options exist these days if we can manage to merge them into Leela Zero. :)

**Comment by roy7** [comment]
I looked back, the paper where I was thinking you had to only use Beta(W+1,L+1) where W and L were integer values came from page 39.3 (page 3 of the PDF) of [Analysis of Thompson Sampling for the Multi-armed Bandit Problem](http://proceedings.mlr.press/v23/agrawal12/agrawal12.pdf).

But they do make the comment:

> We remark that instead of using rt, we could consider more direct and natural updates of type
> Beta(αi, βi) to Beta(αi + ˜rt, βi + 1 − r˜t). However, we do not know how to analyze this because of our essential use of Fact 1, which requires αi, βi to be integral.

So it was because they were working on mathematical proofs they had to do this, not for actual practical implementation reasons.

**Comment by odeint** [comment]
> I do believe stronger search options exist these days if we can manage to merge them into Leela Zero.

For match play that's probably true, but I wouldn't like to abandon the proven self-enhancing cycle of training on the MCTS visit count. It is harder to predict how a different search algorithm would interact with the training cycle/ratchet.

For testing purposes and possibly stronger play with the same networks, it would for sure be nice to have more engine options :-)

**Comment by roy7** [comment]
@odeint Yes it's possible things could diverge too far and need a new experiment and can't be retrofitted to the current experiment.

The thing is, when you learn a policy network from human games and you are solely playing from that policy network (like --visits 1 or a bot like DarkGo), what you are learning is how good the moves are. The highest policy prior is the strongest move. So you can just take the strongest move immediately. In these sorts of setups, you need to learn to read ladders at the nn_eval level so a winning ladder has high prior and a losing ladder has low prior.

In Leela Zero things are somewhat different. What we're learning through the reinforcement step is how good a position is in terms of winning/losing the game (value head/value network in Leela), but we're also learning priors to start the MCTS search where we need it to, in order to increase our chances of winning. I view this differently to the prior example. Using ladders as an example again, Leela Zero learns to play well with ladders not through direct nn_eval but by searching the ladder to the end, seeing if it wins or loses, then propagating that info back up the tree. (The goal of my max_lcb_search branch is to make that propagation instant instead of a slow average-based process.)

When I say "thinks could diverge too far" what I mean is, say we are using a pure Thompson Sampling approach. What we need to know at a board position is what priors to use for the Sampling to speed our finding optimal moves (as discovered during tree search using the value head/value network nn_eval result).

At the end of the day these things may not all be the same thing:
1- Learning the priors of the optimal move (Leela policy network)
2- Learning the priors of the optimal MCTS / UCT1 search starting point (Leela Zero)
3- Learning the priors of the optimal Thompson Sampling starting point (TBD)

I think the ladder example shows a clear difference in #1 and #2. To stop losing ladders, LZ needs to learn super high priors on ladders to search them out to the end.

The question then becomes, can priors be compatible between #2 and #3 enough to allow us to change from one to the other without starting over.

**Comment by Dorus** [comment]
>  Leela Zero learns to play well with ladders not through direct nn_eval but by searching the ladder to the end, seeing if it wins or loses, then propagating that info back up the tree.

This is not true, this is indeed what happens with the policy net during MCTS. However during self play training, eventually the policy net will tell Leela "Play the ladder move", but the value net will tell the MCTS "this is a losing ladder". Right now the value net completely fails here and only tells the MCTS the ladder was failed *after* the entire ladder group is captured.

Worse, because the capture is not required as long as the escaping player does not force the ladder move, the ladder can stay on the board for many more playouts. Forcing the ladder move will get a low score because of the instant capture by the attacking player. Not spending a move on the capture right away, but one move later is going to also benefit the attacking player. Thus Leela tend to be blind to the ladder.

What really should happen is the value net discover the ladder is bad a move before it is captured. More self play games will eventually cause the value net to understand that a ladder is good or bad many moves before, possibly even before the ladder is started. In AG games I've often heard Micheal Redmond claim AG prefers certain moves or plays certain approaches on the other side of the board because a ladder is involved later on. The ladder obvious isn't played after the preparing ladder breaker but it does affect what variation is picked.

LeelaZ so far seems to utterly fail to read ahead in ladders. Possibly it's just a lack of blocks and training time, as ladders did improve a bunch each time we increased net size. However i still believe the value net is to blame here, not the policy net.

**Comment by roy7** [comment]
LeelaZ often reads ladders to the end, those ladders it doesn't fall for. When we see a ladder failure in a game, it's generally because the ladder capture move was invisible. (The prior so low the capture wasn't seen, so ladder didn't end up dying.) The solution to that is higher prior on the ladder capturing move so it's seen in search sooner. Networks that properly see these captures will be stronger than networks that don't. Over time, networks that promote will be slowly moving in the direction of high enough priors to never fail to read a ladder properly, and thus never lose by playing a broken ladder.

If the value network ever learns this well enough that early steps in the ladder are given low win rates, then MCTS will change sooner, the priors for training will adjust (since less plays were observed going into those moves), and LZ will stop playing the ladders out. It'll be self correction if the value network ever really does learn it.

**Comment by roy7** [comment]
I want to do some searching on setting priors in Thompson Sampling contexts. But I have plans tonight and tomorrow morning, so it'll be later this weekend. Work is a bit busy today. I did find one paper of general interest perhaps, On the [Prior Sensitivity of Thompson Sampling](https://arxiv.org/abs/1506.03378) where they prove the regret bounds for good and bad priors.

**Comment by roy7** [comment]
[Racing Thompson: an Efficient Algorithm for Thompson Sampling with Non-conjugate Priors](https://pdfs.semanticscholar.org/429f/ec5ffc146d4fe339caa9e27015e507d4e26c.pdf)

**Comment by tterava** [comment]
I'm running ringmaster with roy7's LCB-pull request, gcp's Thompson-branch and regular next-branch. This way we'll have a better understanding how each of these implementations does on the latest 192x15 network.

I've set it up to do 500 games each matchup with -v 1600, so it should take 2-3 days to finish the experiment.

Command line: `"leelaz -g -t 11 -d -q -v 1600 -r 5 --gpu 0 --timemanage off --noponder -w leelaz-model-swa-8-64000.txt", startup_gtp_commands=["time_settings 0 101 1"]`

This should also be a good pressure test for the latest threading fixes. My setup has been extremely sensitive to any race conditions. I could barely generate 10 moves without LZ halting mid-thinking.

**Comment by tterava** [comment]
Thompson branch is currently 0-75 against Next and 3-72 against LCB. There's probably an issue with the implementation so I'm dropping it out.

**Comment by Eddh** [comment]
Yeah max_lcb_root seems to be like a very good idea. It is close to what we have since more visits means a higher lower bound. Except it can compensate for all those unfortunate cases where the move with the most visit is actually found to be worse than another, but the search then need to spend twice as much playouts to really switch.

**Comment by tterava** [comment]
Yeah LZ keeps track of time by default, so it's theoretically possible that late game the -v 1600 cap won't be reached because of time control. I wanted to rule out that possibility.

Here's my current results. I've actually started to think that LCB makes a difference regardless of visit count. It's definitely great with high visit counts but it seems to be very good even with low count. It doesn't seem to have any downsides.

Here's progress so far:

LCB v Next (149/500 games)
board size: 19   komi: 7.5

------------ wins ------------ black ------------- white -------- avg cpu       
LCB --- 90 60.40% ----- 44 58.67% ----- 46 62.16% --- 258.57
Next --- 59 39.60% ----- 28 37.84% ----- 31 41.33% --- 260.68
------------------------------- 72 48.32% ----- 77 51.68%

**Comment by Eddh** [comment]
I used to think it would not make a difference at high visit count, but it's actually still useful because even after tens of thousands of playouts, the best value estimation move can change, and then it will have to catch up in visits, unless you use lcb.

**Comment by tterava** [comment]
I stopped the test a bit early because the results are unambigious: LCB is the clear winner.
Here's the results:
```
LCB v Next (400/400 games)
board size: 19   komi: 7.5


            wins              black          white        avg cpu
LCB       249 62.25%       124 62.00%     125 62.50%      258.38
Next      151 37.75%       75  37.50%     76  38.00%      259.55
                           199 49.75%     201 50.25%
```

The boost in performance is so significant that I would almost consider bumping the version number on this change alone, as it will instantly produce a lot higher quality games with seemingly no downsides. I noticed on the pull request there was some data which suggests --timemanage should still be turned on, even if it doesn't really make sense considering how LCB interacts with it. Maybe next step would be to make a smart time management scheme that works better with this. Perhaps making a move early if some confidence bounds are reached?

**Comment by Dorus** [comment]
Quick Q: how do you plan on incorporating LCB in training? Should training be done on LCB value of still tick to playouts? As far as i understood LCB only changes move selection and not playouts right?

**Comment by tterava** [comment]
This implementation only changes how the move is selected. Network still produces playouts as usual, but we don't simply pick the move that was explored the most. There are probably even better ways to choose the move or direct the actual search, but I think this is a great first step.

In short: No other changes should be required. Free ELO.

**Comment by Eddh** [comment]
I think the value net probably makes some estimate of what move will be played next to better predict winrate. It will just have to adjust to this new move picking method, which will be interesting.

**Comment by roy7** [comment]
Right, I don't think any training changes are needed. Ultimately training is about specific board states and the MCTS result proportions of search on each move, with a W or L final result for the value head to learn from. Regardless of the move chosen that stays the same. However, the final result of the match (W or L) may change since actual play is stronger but it's equal for both players in selfplay, so the hope is it'd give us more accurate W or L data by playing stronger. (And of course actual play, LZ would be stronger.)

Thanks for the tests @tterava. I had speculated that timemanage might undo the benefits of max_lcb_root, but that concern is apparently unimportant because our failsafe behavior is just normal /next behavior anyway. Thinking longer might find more max_lcb_root move changes, but from @OmnipotentEntity's tests here https://github.com/gcp/leela-zero/pull/883#issuecomment-382768285 we know absolute short clocks, timemanage is much more important than max_lcb_root.

@OmnipotentEntity is doing more tests, and will also confirm next+timemanage is much stronger than next-timemanage in 7.5 minute absolute time (his first test I think came down to that, nothing about max_lcb_root). He'll also try some scenarios with byoyomi time.

**Comment by roy7** [comment]
@gcp Why go with a beta_prior of 15 instead of 1? 15 gives fairly steep curve around .50 win rate, where as 1 is a uniform "no info known" prior typically used by the literature. (I admit I just started looking over your code so may be misunderstanding the use of that default value.)

**Comment by gcp** [comment]
>@gcp Why go with a beta_prior of 15 instead of 1?

It's tuned for strenght. (But as illustrated, still very weak). I'm trying a different implementation, will upload it tomorrow. I didn't compare any of this to next, just checked if it behaves reasonably.

**Comment by roy7** [comment]
Ok well having the prior based on assuming a 50% win rate with 16 wins/16 losses (basically) would give us the sort of "depth first" when we are winning and "width first" when we are losing that we saw during the fpu experiments. My concern as always is that over confidence (very high win rate when we're actually in danger) might result in too little exploration to avoid a blunder?

But going with uniform prior (1,1) means a lot of search at each node which is fine for theoretical papers and other games, but our branching factor isn't practical for it. (Although with heatmap values low enough those children never get search anyway.)

**Comment by roy7** [comment]
I tested existing Thompson Sampling branch at game start with 1 and 15 priors, to compare difference in branch/search depth. 3200 visits. Was just curious.

Beta prior 1:

5.7 average depth, 18 max depth
1588 non leaf nodes, 2.02 average children
3202 visits, 1144111 nodes, 3201 playouts, 488 n/s

5.7 average depth, 17 max depth
1588 non leaf nodes, 2.02 average children
3202 visits, 1144111 nodes, 3201 playouts, 488 n/s

5.7 average depth, 17 max depth
1578 non leaf nodes, 2.03 average children
3203 visits, 1144478 nodes, 3202 playouts, 372 n/s

Beta prior 15:

9.0 average depth, 29 max depth
2043 non leaf nodes, 1.57 average children
3202 visits, 1133400 nodes, 3201 playouts, 541 n/s

9.1 average depth, 32 max depth
2012 non leaf nodes, 1.59 average children
3204 visits, 1133875 nodes, 3203 playouts, 431 n/s

9.0 average depth, 27 max depth
2035 non leaf nodes, 1.57 average children
3201 visits, 1133081 nodes, 3200 playouts, 377 n/s

**Comment by roy7** [comment]
Searching around for discussions of priors, [A Tutorial on Thompson Sampling](https://arxiv.org/pdf/1707.02038.pdf) on page 21 uses an example of web based advertising, where prior ads have various click-through rates. By clustering the prior ads into various groups, and the seeing which group the new ad is more similar to, you could use the click rate of the group as a prior for the new ad by setting a Beta prior similar to the group's prior empirical results.

I imagine constructing useful priors is so domain-specific there won't be any sort of general published solution for it?

At some point FPU was using the parent's eval as a prior. In effect that means "search deeper if our win rate has risen" or "search wider if our win rate has fallen".

**Comment by roy7** [comment]
@punter11235 Note the win rate estimate when reading out the ladder is 3%. It might be an issue that all moves look so bad it's hard to find something else worth playing.

**Comment by punter11235** [comment]
Well, other moves have 1.8%. I realize it might be an issue but still, going into 0% line instead of 1.8% line is quite bad.

**Comment by isty2e** [comment]
It can be interesting to see what lcb_root branch plays in this situation.

**Comment by tterava** [comment]
It's also possible the network has learnt absolute minimum it needs to play a fuseki where it doesn't fall behind. That leaves more capacity in the network to learn more important things, like mid game and late game strategy and tactics. However the assumptions the network makes are only valid in even game.

It has never been in a situation where there's more black stones on the board so early in the game so i would expect all kinds of weird weaknesses to arise.

**Comment by roy7** [comment]
@gcp Thompson Sampling experiment shows what I mean sometimes about the math not quite fitting with our application. Each pull of the bandit we make isn't another sample from one distribution, it's a nn_eval mean from a different distribution with unknown variance/etc. So maybe tons of arm pulls are Normal under Central Limit Theorem, I don't know. (@kiudee?)

Using a NN where we learn distributions for the value head instead of just a mean (so we can construct a confidence bound on the result) would give us useful info for a node even with just 1 nn_eval. Not for LZ but maybe some fork or future experiment. :)

**Comment by punter11235** [comment]
>  In normal play the winrate difference is still high enough that the ladder isn't played, but when all moves are <3% winrate this happens.

Ok, here is an example with ~20% winrate for the ladder and 10%+ for other moves (2 stones handicap)
![ladder5](https://user-images.githubusercontent.com/508706/39115976-7ba0c776-46e3-11e8-9828-4d60e6c98e22.jpg)

> It has never been in a situation where there's more black stones on the board so early in the game so i would expect all kinds of weird weaknesses to arise.

It's probably that. I was unable to construct a similar situation without a handicap.

**Comment by punter11235** [comment]
> In normal play the winrate difference is still high enough that the ladder isn't played, but when all moves are <3% winrate this happens.

>It has never been in a situation where there's more black stones on the board so early in the game so i would expect all kinds of weird weaknesses to arise.

Here is an example from normal play (all moves at least reasonable according to Leela up to this point) where 500k+ playouts i still not enough to read the ladder. Other alternatives sit at 40-44%+ winrate:

![ladder10](https://user-images.githubusercontent.com/508706/39119931-69df1da8-46ed-11e8-9af1-0dd8048e5783.jpg)

If the search never punish her for extending the ladder she will never learn...

**Comment by bood** [comment]
could you upload the sgf?

punter11235 <notifications@github.com> 于 2018年4月23日周一 17:57写道：

**Comment by gcp** [comment]
I pushed an update: https://github.com/gcp/leela-zero/tree/thompson_sampling

This one uses a formula similar to our FPU reduction to turn the policy prior into a modification of the estimated distribution mean, and then fiddles with the prior (and only the prior).

It's much stronger than my previous attempt, but probably still a lot weaker than UCB1.

**Comment by punter11235** [comment]
> could you upload the sgf?
I have found more convincing example from 2 handicap game:
![ladder5](https://user-images.githubusercontent.com/508706/39172584-7902a88a-47a2-11e8-8e3b-50e3f7d5cfaf.jpg)

SGF:
https://pastebin.com/zdmjJN9m

I was unable to find a similar example in even game (even if I add a useless move like capturing bottom left or even a stone on 1st line it starts leading the ladder correctly).

**Comment by gcp** [comment]
```
leelaz-next v leelaz-th2 (33/100 games)
board size: 19   komi: 7.5
              wins              black         white       avg cpu
leelaz-next     28 84.85%       14 82.35%     14 87.50%    164.99
leelaz-th2       5 15.15%       2  12.50%     3  17.65%    220.58
                                16 48.48%     17 51.52%
```

The Thompson branch isn't a total loss, though a lot weaker. I might try some more variations.

**Comment by gcp** [comment]
@roy7 this is somewhat similar to what the latest push has, but instead of tweaking the confidence based on the FPU with policy adjustment, it adjusts the estimated mean.

**Comment by roy7** [comment]
For what it's worth, opening game state with latest copy of the Thompson Sampling branch after @gcp's changes (updating https://github.com/gcp/leela-zero/issues/860#issuecomment-383400848):

5.2 average depth, 12 max depth
1611 non leaf nodes, 1.99 average children
3201 visits, 1145259 nodes, 3200 playouts, 1265 n/s

5.2 average depth, 11 max depth
1629 non leaf nodes, 1.96 average children
3201 visits, 1145255 nodes, 3200 playouts, 1077 n/s

5.2 average depth, 10 max depth
1601 non leaf nodes, 2.00 average children
3201 visits, 1145314 nodes, 3200 playouts, 1285 n/s

I didn't test later actual game positions, but the opening search is less deep on average and about half as deep in max depth.

For comparison, current /next client:

6.2 average depth, 19 max depth
2219 non leaf nodes, 1.44 average children
3201 visits, 1142023 nodes, 3200 playouts, 1123 n/s

Just a random observation that TS is searching a bit wider at game start than UCT1.

**Comment by roy7** [comment]
@gcp It seems current TS branch has a lot of extra stuff in the algorithm like adjusting the beta priors based on psa / cfg_fpu_reduction, sqrt(policy_ratio), etc. Is that just left over UCT1 stuff or intentionally for the TS testing? The cfg_puct = .1 also means the wins/losses are being magnified 10x when creating the distributions. I'd think the low win rate choices would be excluded rapidly but policy_eval must be offsetting that, since I still see low prior moves get lots of visits at game opening.

**Comment by gcp** [comment]
>But isn't our prior just a "better than nothing" value we have to use instead of using the Beta(1,1) non-informative prior? In UCT 1 might use parent eval or parent-.2 or whatever, but it's all an arbitrary way to try and adjust exploration?

You are confusing the policy prior with the value estimation.

In UCT1 the policy prior is a very long term bias towards moves that the network likes, which is quite reasonable as it statically predicts the search move after 3200 visits with >64% accuracy these days...

**Comment by gcp** [comment]
> intentionally for the TS testing

Intentional.

>policy_eval must be offsetting that, since I still see low prior moves get lots of visits at game opening.

That's due to the cfg_beta_prior vs cfg_fpu_reduction. Small fpu_reduction and large beta prior means moves are not immediately thrown away after a few losses. (Nor overly liked after a few wins)

These were all tuned for strength.

**Comment by alreadydone** [comment]
Has anyone ever tuned the softmax temperature (for the policy prior) for strength? It's unclear to me that t=1 is optimal.

**Comment by tterava** [comment]
I did additional testing with @roy7 's implementations and here are all the results i got:

-t 11 and -v 1600:
```
                 wins             black          white
LCB_root       249 62.25%       124 62.00%     125 62.50%
Next           151 37.75%       75  37.50%     76  38.00%
```

-t 1 and -v 1600 (LCB_full isn't thread safe yet):
```
LCB_full vs LCB_root: 194 - 192 (50.26 %)
LCB_full vs Next: 221 - 181 (54.98 %)
```

-t and -v 200:
```
              wins            black           white
lcb_root    56 70.89%       28 70.00%       28 71.79%
next        23 29.11%       12 30.00%       11 28.21%
                            40 50.63%       39 49.37%


              wins            black           white
lcb_full   227 49.35%      105 49.30%      122 49.39%
lcb_root   233 50.65%      108 50.70%      125 50.61%
                           213 46.30%      247 53.70%

              wins            black           white
lcb_full    64 68.09%       31 68.89%       33 67.35%
next        30 31.91%       14 31.11%       16 32.65%
                            45 47.87%       49 52.13%
```

LCB_full seems to be at least as strong as Next in all the situations tested, but it's inconclusive whether it's stronger than LCB_root. Its winrate against Next doesn't seem as high, but it might be because I couldn't test it with 11 threads. I was told that increasing thread count makes the search generally wider, and that's where most gains seem to be had with the LCB approach.

**Comment by roy7** [comment]
I find the low visit count results fascinating. I was always stuck only thinking about move 1 has tons of visits but move 2 is better and there's no time for move 2 to over take move 1. And as a result, also thinking this would only be useful with higher visit counts where there's time for strong confidence bounds/etc.

But the -v 200 results show even with few visits the concept is sound and useful. For my own education I should run a -v 200 game with _root and look over the log to see all the times it overrode the max visit move, to better understand what is happening.

I'm both happy and sad _full never does much better than _root. _full does a lot more than _root does, yet sometime _root always evenly matches with it and _root is so naively simple by comparison.

**Comment by roy7** [comment]
@gcp Is choosing the move with the highest # of visits still correct move selection in the Thompson Sampling branch? If so I wonder how max_lcb_root merged with thompson_sampling would do. TS searches wider, and max_lcb_root would perhaps make other choices than "most visits" more often.

**Comment by Mardak** [comment]
@bood provided a ladder game to analyze, and I used elf 224x20 to analyze with leelaz using just symmetry 0 and with lcb and with thompson sampling:

<details><summary>sgf details</summary><p>

![image](https://user-images.githubusercontent.com/438537/39667792-447be8f0-5072-11e8-8b8a-6450f34dc041.png)
```sgf
(;GM[1]FF[4]RU[Chinese]DT[2018-05-05]SZ[19]KM[7.5]TM[3600]PB[Leela Zero 0.14 192x15]PW[Human]RE[W+7.5]
C[Leela Zero options: -w ../192x15 --noponder --seed 5922048003111841579]

;B[pp];W[pd];B[cp];W[dd];B[cf];W[fc];B[fp];W[qq];B[qp];W[pq]
;B[nq];W[oq];B[op];W[nr];B[rq];W[rr];B[or];W[rp];B[qr];W[sq]
;B[pr];W[rq];B[mr];W[np];B[mq];W[qn];B[no];W[mp])
```
</p></details>

The play for black move 29 should be M4.

With regular search and 100,000 visits, it finds it after spending a lot of visits with Q6:
```
  M4 ->   79831 (V: 87.72%) (N:  7.95%) PV: M4 C11 C9 D3 D4 F3 E3 E2 E4 G2 G3 F2 D2 H3 G4 K3 K5 D9 C10 D10 D11 C12 D12 C8 B8 D13 C13 B11 B10 A10 B9 B13 A12 B12 E13 B14
  Q6 ->   20139 (V: 51.34%) (N: 88.32%) PV: Q6 N5 R7 S6 Q7 L4 K3 S7 R9 S8 N7 L7 L3 L9 N9 C11 C9 E11 E9 E3 E4 C3
```

Only after visiting Q6 20,139 times does it decide to explore M4 more, and with LCB #883, it would choose M4 with only needing a little bit more visits (whereas normal search would require ~20k more visits to switch, so total ~40k visits):
```
20,200 visits just before search decides to switch to M4
  Q6 ->   20102 (V: 51.34%) (N: 88.32%) (LCB: 50.25%) (UCB: 52.44%) PV: Q6 N5 R7 S6 Q7 L4 K3 S7 R9 S8 N7 L7 L3 L9 N9 C11 C9 E11 E9 E3 E4 C3
  M4 ->      68 (V: 38.74%) (N:  7.95%) (LCB: 21.60%) (UCB: 58.10%) PV: M4 N5 Q6 O6 Q7 R7 Q8 M5 L5 L6 L4

20,300 visits with rapidly rising win rate and LCB
  Q6 ->   20139 (V: 51.34%) (N: 88.32%) (LCB: 50.25%) (UCB: 52.43%) PV: Q6 N5 R7 S6 Q7 L4 K3 S7 R9 S8 N7 L7 L3 L9 N9 C11 C9 E11 E9 E3 E4 C3
  M4 ->     131 (V: 61.63%) (N:  7.95%) (LCB: 47.78%) (UCB: 74.26%) PV: M4 N5 N6 M5 L5 M6 M7 L6 K6 L7 L8 K7 J7 K8 K9 J8 H8 J9 J10 H9 G9 H10 H11 G10 F10 G11 G12 F11 E11 F12 F13 E12 D12 E13 E14 D13 C13 D11 D14 T5 S1 T1

20,400 visits to switch to M4
  M4 ->     231 (V: 77.56%) (N:  7.95%) (LCB: 68.12%) (UCB: 85.36%) PV: M4 N5 N6 M5 L5 M6 M7 L6 K6 L7 L8 K7 J7 K8 K9 J8 H8 J9 J10 H9 G9 H10 H11 G10 F10 G11 G12 F11 E11 F12 F13 E12 D12 E13 E14 D13 C13 D11 D14 T5 S1 T1 R1 P1 O1
  Q6 ->   20139 (V: 51.34%) (N: 88.32%) (LCB: 50.25%) (UCB: 52.43%) PV: Q6 N5 R7 S6 Q7 L4 K3 S7 R9 S8 N7 L7 L3 L9 N9 C11 C9 E11 E9 E3 E4 C3
```

And with Thompson Sampling https://github.com/gcp/leela-zero/tree/thompson_sampling, fewer than 100 visits to decide on M4:
```
  M4 ->      81 (V: 76.03%) (N:  7.95%) PV: M4 N5 N6 M5 L5 M6 M7 L6 K6 L7 L8 K7 J7 K8 K9 J8 H8 J9 J10 H9 G9 H10 H11 G10
  Q6 ->      18 (V: 40.44%) (N: 88.32%) PV: Q6 R7 M4 N5
```

**Comment by roy7** [comment]
Although we treat the rewards of a move (bandit arm) as if they don't change when doing Thomson Sampling, the truth is the mean is always moving around as the search goes deeper and if we find specifically better or worse moves, the true average of the move will be radically different than what we've collected so far (what I try to address in max_lcb_full branch). I found a paper called [Thompson Sampling in Switching Environments with Bayesian Online Change Point Detection](https://arxiv.org/pdf/1302.3721.pdf) which sort of addresses this by detecting when changes to the bandit arm(s) have happened.

**Comment by roy7** [comment]
Speaking of which, here's a paper on UCBT with arms that have changing rewards and overcoming UCBT inertia (the whole point of max_lcb_root).

[Change Point Detection and Meta-Bandits for Online Learning in Dynamic Environments](https://www.researchgate.net/publication/29646788_Change_Point_Detection_and_Meta-Bandits_for_Online_Learning_in_Dynamic_Environments).

It's the source paper for an algorithm called Adapt-EvE which is referred to in prior post I made.

**Comment by roy7** [comment]
@gcp Since Thompson Sampling goes so wide, and we have so many moves and a high branching factor... an idea you might try is prune all moves with a policy network of .01 or lower? Or even lower than that. But like at game start, we don't need to sample all 361 moves. The heatmap only shows 16 possible moves we should be considering, the rest being rounded to 0.

 So instead of trying to use policy to adjust the Beta() of the moves, instead we'll use policy to prune the search away from super unlikely moves.

**Comment by roy7** [comment]
A few interesting papers.

[Learning to Optimize Via Information-Directed Sampling](https://arxiv.org/pdf/1403.5556.pdf)
[Thompson Sampling for Complex Online Problems](http://proceedings.mlr.press/v32/gopalan14.pdf)
[Thompson Sampling for Monte Carlo Tree Search and Maxi-min Action Identification](https://www.universiteitleiden.nl/binaries/content/assets/science/mi/scripties/master/2017-2018/master-thesis_mingxi-li_12okt2017.pdf)

**Comment by roy7** [comment]
[An Empirical Evaluation of Thompson Sampling](https://papers.nips.cc/paper/4321-an-empirical-evaluation-of-thompson-sampling.pdf) has some discussion of use of priors.

[Prior-free and prior-dependent regret bounds for Thompson Sampling](https://arxiv.org/pdf/1304.5758.pdf)

**Comment by Videodr0me** [comment]
I did a number of mcts search experiments on leela zero chess (lc0). KillerDucky asked me to cross-post this here as some of it might be relevant for Go.
https://github.com/Videodr0me/leela-chess-experimental

**Comment by roy7** [comment]
I didn't realize it at first, but the core of the last paper in my prior post says:

> Next, the algorithm proceeds to remove the arms whose upper bounds of estimated expected reward are less than the lower bound of the current best arm.

Which was the "full max_lcb" idea step to follow after https://github.com/gcp/leela-zero/pull/883.

**Comment by MaurizioDeLeo** [comment]
@roy7 What I gather from this issue is that your branch "LCB-root" is 100 elo stronger than the master.
Any reason why this is not imported back into gcp-master ?

**Comment by tterava** [comment]
@MaurizioDeLeo The results are inconsistent. It's been tested quite extensively and it might be stronger on some systems and engine parameters. However it's not clear whether it provides any benefit overall.

**Comment by roy7** [comment]
@MaurizioDeLeo It was reliable before the timemanagement feature was added. But with timemanagement, the results were less consistent. But timemanagement without max_lcb was stronger than max_lcb without timemanagement. So timemanagement is in and max_lcb a fond memory I do hope to experiment with more in the future. :)

**Comment by MaurizioDeLeo** [comment]
@roy7 , any chance to have *both* LCB and time management ?
Maybe the improvements are additive ?

Maybe instead of stopping when a move can not be overcome in term of visit, we can stop based on some other criteria ?

**Comment by roy7** [comment]
Yeah it's possible in theory. I was just thinking about it the other night, but I'm not sure what stopping criteria would make sense.

**Comment by roy7** [comment]
Was scrolling down the list of Deepmind papers and found this interesting one. [Garbage In, Reward Out: Bootstrapping Exploration in Multi-Armed Bandits](https://arxiv.org/abs/1811.05154v1)

> We propose a multi-armed bandit algorithm that explores based on randomizing its history. The key idea is to estimate the value of the arm from the bootstrap sample of its history, where we add pseudo observations after each pull of the arm. The pseudo observations seem to be harmful. But on the contrary, they guarantee that the bootstrap sample is optimistic with a high probability. Because of this, we call our algorithm Giro, which is an abbreviation for garbage in, reward out. We analyze Giro in a K-armed Bernoulli bandit and prove a O(KΔ−1logn) bound on its n-round regret, where Δ denotes the difference in the expected rewards of the optimal and best suboptimal arms. The main advantage of our exploration strategy is that it can be applied to any reward function generalization, such as neural networks. We evaluate Giro and its contextual variant on multiple synthetic and real-world problems, and observe that Giro is comparable to or better than state-of-the-art algorithms.

**Comment by roy7** [comment]
kiudee shared a couple new papers with me:

[Sequential Test for the Lowest Mean: From Thompson to Murphy Sampling](http://papers.nips.cc/paper/7870-sequential-test-for-the-lowest-mean-from-thompson-to-murphy-sampling)
[Monte-Carlo Tree Search by Best Arm Identification](http://papers.nips.cc/paper/7075-monte-carlo-tree-search-by-best-arm-identification)

**Comment by roy7** [comment]
New Deepmind/Google paper. 

[Empirical Bayes Regret Minimization](https://arxiv.org/abs/1904.02664)

> The prevalent approach to bandit algorithm design is to have a low-regret algorithm by design. While celebrated, this approach is often conservative because it ignores many intricate properties of actual problem instances. In this work, we pioneer the idea of minimizing an empirical approximation to the Bayes regret, the expected regret with respect to a distribution over problems. This approach can be viewed as an instance of learning-to-learn, it is conceptually straightforward, and easy to implement. We conduct a comprehensive empirical study of empirical Bayes regret minimization in a wide range of bandit problems, from Bernoulli bandits to structured problems, such as generalized linear and Gaussian process bandits. We report significant improvements over state-of-the-art bandit algorithms, often by an order of magnitude, by simply optimizing over a sample from the distribution.

**Comment by iopq** [comment]
So the first problem is choosing the best move from A and B.

We have some statistics available, like N of visits, the evaluation itself and some measure of uncertainty. We used to pick the highest N, but we came to the conclusion that we pick better lines with max LCB as long as N > 10% of all visits.

But this means our search is NOT actually doing what it should when it knows it has N visits. There should be three things it should seek to do:

1. Get a second move 10% visits and keep it there if there's a chance it has the highest win rate. No point in searching the main line when it's increasingly unlikely to change the value after a lot of visits.
2. Spread visits to low prior moves when the top few options don't yield promising candidate moves. We do this VERY poorly, sometimes it takes 300,000 visits to find the sharp line when the prior is low for it.
3. When dumping visits to another move since it seems more promising, also give a third candidate up to 10% visits in case the alternative move doesn't do anything. And so on, I expect a lot of moves to get to the 10% mark in case we find a refutation for the current main line. That way we have alternatives.

The second problem is back propagating the values from the nodes we have. We currently average. We can try the "mix" strategy of also including the value of the best node as per Remi Coulom. This should sharply drop the value of a losing line or quickly improve a sharp tesuji line once found. This is actually not a problem at a lot of visits, but at 1600 it doesn't have time to earn win rate quickly enough if found at the end of the search.

**Comment by roy7** [comment]
[INFINITE ARMS BANDIT: OPTIMALITY VIA CONFIDENCE BOUNDS](https://arxiv.org/pdf/1805.11793.pdf)

> The infinite arms bandit problem was initiated by Berry et al.(1997). They derived a regret lower bound of all solutions for Bernoulli rewards with uniform priors, and proposed bandit strategies based on success runs, but which do not achieve this bound. Bonald and Proutiere (2013) showed that the lower bound was achieved by their two-target algorithm, and extended optimality to Bernoulli rewards with general priors. We propose here a confidence bound target (CBT) algorithm that achieves optimality for general, unknown reward distributions. For each arm we require the mean and standard deviation of its rewards to compute a confidence bound. We play the arm with the smallest bound provided it is smaller than a target mean. If the bounds are all larger, then we play a new arm. We show how for a given prior, the target mean can be computed to achieve optimality. In the absence of information on the prior, the target mean can be determined empirically, and we show that the regret achieved is still comparable to the regret lower bound. Numerical studies show that CBT is versatile and outperforms its competitors.
